mtls_server/auth.py

```diff
@@ -0,0 +1,93 @@
+import base64
+import os
+import time
+import pprint
+import json
+from functools import wraps
+
+from flask import request
+from flask import g
+from flask import current_app
+
+from .logger import logger
+from .utils import error_response
+from .utils import time_in_range
+from .utils import write_sig_to_file
+
+
+pp = pprint.PrettyPrinter(indent=4)
+
+
+class SignatureTimestampOutOfBoundsException(Exception):
+    pass
+
+
+class MissingTokenException(Exception):
+    pass
+
+
+class BadSignatureException(Exception):
+    pass
+
+
+def login_required(f):
+    @wraps(f)
+    def login_required_wrap(*args, **kwargs):
+        allowed_tokens = ['PGP-SIG']
+        g.is_admin = False
+
+        if not request.headers['Authorization']:
+            return error_response("authentication required", 401)
+
+        token_type, token = request.headers['Authorization'].split(' ')
+
+        if token_type not in allowed_tokens:
+            return error_response("authentication required", 401)
+
+        if token_type == "PGP-SIG":
+            data = request.get_data()
+
+            # Move the pointer to the input stream back to 0 so that this can be accessed
+            # by the actual request
+            request.input_stream.seek(0)
```

Comment:
Wait..... so if the endpoint has "login required", then you pass a PGP signature as the GET/POST body?
I'm not a fan of that design

Reply:
No, it is passed as a Authorization header and is b64encoded. The reason there is a seek in that call is because we need to be able to verify the content of the body with the detached signature.

The original design had the signature in the body and generally only signed part of the data. This now signs all of the data but does have the fun problem that all fields in the json object must now be alphabetized to to allow for verification since json.loads has a sort_keys attribute and does so by default in flask apparently